File: apps/android/app/src/main/kotlin/app/storyarc/AppShell.kt
package app.storyarc

import android.net.Uri
import androidx.activity.ComponentActivity
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.MenuBook
import androidx.compose.material.icons.filled.Download
import androidx.compose.material.icons.filled.Home
import androidx.compose.material.icons.filled.Inventory2
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.Settings
import androidx.compose.runtime.Composable
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.viewmodel.compose.viewModel
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import app.storyarc.core.designsystem.back.PredictiveBack
import app.storyarc.core.designsystem.back.PredictiveBackHost
import app.storyarc.core.designsystem.navigation.AdaptiveNavigationShell
import app.storyarc.core.designsystem.navigation.NavigationEntry
import app.storyarc.core.designsystem.navigation.RailMenuLabels
import app.storyarc.core.model.AppSettings
import app.storyarc.core.model.PublicationFormat
import app.storyarc.core.model.QuickActionRequest
import app.storyarc.core.model.SourceConnectionState
import app.storyarc.core.persistence.RemovedDownload
import app.storyarc.feature.epubreader.EpubReaderActivity
import app.storyarc.feature.library.CataloguePage
import app.storyarc.feature.library.KavitaLevel
import app.storyarc.feature.library.KavitaPage
import app.storyarc.feature.library.LibraryViewModel
import app.storyarc.feature.library.SmbPage
import app.storyarc.feature.settings.BuildInfo
import app.storyarc.feature.settings.WhatsNew
import app.storyarc.feature.settings.WhatsNewRelease
import app.storyarc.feature.settings.WhatsNewSheet
import app.storyarc.feature.settings.WhatsNewStore
import app.storyarc.navigation.AppDestination
import app.storyarc.navigation.AppNavigation
import app.storyarc.navigation.AppSheet
import app.storyarc.navigation.Screen

/**
 * The frame the whole app draws inside.
 *
 * Three destinations in Material's own adaptive navigation, one typed stack of screens per
 * destination, and **one** back rule. What this replaces was a chain of fourteen booleans
 * and nullables in which each branch installed its own `BackHandler` — so "what does back
 * do" had fourteen answers that had to agree, and a rail whose selected item was re-derived
 * from whichever of four nullables happened to be set.
 *
 * **Search is the fourth destination**, and this comment used to say the opposite: that
 * Material ranks a search bar above a search destination and permits the destination only
 * for an app whose primary action is searching. Material does say that, and the judgement it
 * conditions is ours — StoryArc's publications arrive from a device, a folder, an OPDS
 * catalogue, a Kavita server and an SMB share, and no shelf shows all of them at once in a
 * way a reader can scan. Search is the only surface that spans the sources, so in this app it
 * is the way in rather than a filter. `quiet-shell-and-search`'s `design.md` sets the
 * argument out in full.
 */
@Composable
internal fun AppShell(
    activity: ComponentActivity,
    dependencies: AppDependencies,
    settings: AppSettings,
    onSettingsChange: (AppSettings) -> Unit,
    onResetSettings: () -> Unit,
    /** A file the system handed over, waiting for the composition to pick it up. */
    handedOver: MutableState<Uri?>,
    /** A quick action the launcher sent, waiting for the same. */
    quickAction: MutableState<QuickActionRequest?>,
) {
    var navigation by rememberSaveable(stateSaver = AppNavigation.Saver) {
        mutableStateOf(AppNavigation())
    }
    var sheet by remember { mutableStateOf<AppSheet?>(null) }

    // What changed in the version just installed, if this launch is the one that says so.
    //
    // `settings-and-about`: shown "once, after it has been updated", never on a first ever
    // launch, and never twice for one version. All three are `WhatsNew.onLaunch`'s answer,
    // and it is asked **inside `remember`'s initial value**: the version is recorded by the
    // same call that decides, so a reader who swipes the sheet away — or one who never sees
    // it because the version had nothing to say — is recorded either way. There is no path
    // through this that shows the screen without recording it, which is the requirement
    // stated as "the seen flag is written when the screen is shown, not when it is
    // dismissed".
    var whatsNew by remember {
        mutableStateOf<WhatsNewRelease?>(
            WhatsNew.onLaunch(
                installed = BuildInfo.version,
                store = WhatsNewStore.open(activity.applicationContext),
            ),
        )
    }
    var refusedFile by remember { mutableStateOf<OpenedFile.Outcome?>(null) }

    // Re-read on the way in, so a download made while browsing an online library is on this
    // screen rather than one launch behind it.
    val downloads = remember { mutableStateOf(dependencies.downloads.library()) }
    val removed = remember { mutableStateOf<RemovedDownload?>(null) }

    // Held across every destination, not inside the library's: the reader's end screen asks
    // it what comes next in the series, and a model created inside one branch would not
    // exist to ask.
    val library = viewModel<LibraryViewModel>(
        factory = viewModelFactory {
            initializer {
                LibraryViewModel(
                    activity.application,
                    dependencies.progress,
                    dependencies.libraryPreferences,
                    dependencies.sources,
                    dependencies.shelves,
                    // One store, two readers of it: what was downloaded joins the one
                    // library rather than being reachable only by browsing back to the
                    // server it came from, and imported copies live beside it.
                    dependencies.downloads,
                    dependencies.scanJournal,
                    dependencies.kavitaCards,
                )
            }
        },
    )

    val host = AppHost(
        activity = activity,
        dependencies = dependencies,
        library = library,
        downloads = downloads,
        removed = removed,
        navigate = { navigation = it(navigation) },
        open = { publication, path ->
            // `native-experience` asks for continuity. Android has no Handoff and there is
            // no backend to invent one with, so the honest mirror is to tell the system:
            // the entry is pushed *and* reported as used, which is what lets the launcher
            // rank it and the Assistant answer for it. Here rather than on a timer, because
            // opening a book is the event.
            HomeScreenActions.reportOpened(
                activity,
                publication,
                downloads.value.downloads.isNotEmpty(),
            )
            // Two readers, chosen by what the publication *is* rather than by a mode the
            // reader picks. A reflowable book is laid out by a rendering engine (ADR-0005);
            // a comic is a list of images and needs none. A fixed-layout EPUB is the third
            // case and belongs with the comic reader.
            if (publication.format == PublicationFormat.EPUB && !publication.isFixedLayout) {
                activity.startActivity(
                    EpubReaderActivity.intent(
                        activity,
                        path,
                        publication.displayTitle,
                        publication.series,
                    ),
                )
            } else {
                // Replaced rather than stacked when a reader is already open: the next
                // volume offered at the end of one would otherwise leave a pile of readers
                // behind a long series.
                val reader = Screen.Reader(publication, path)
                navigation = if (navigation.current is Screen.Reader) {
                    navigation.replace(reader)
                } else {
                    navigation.push(reader)
                }
            }
        },
        openPage = { publication -> navigation = navigation.openPage(publication) },
        browse = { source, term ->
            // One tap, three kinds of place, decided by what the source is. The reader
            // picked somewhere to browse, not a protocol.
            navigation = when {
                // A source that is not answering is not opened into a browser that will
                // fail: it would fetch, wait, and land on an empty list with nothing to say
                // why. `sources` promises "cached contents remain browsable", and for a
                // server there are none — saying so is the honest thing besides.
                source.state is SourceConnectionState.Unreachable ->
                    navigation.push(Screen.SourceAway(source))
                // `kavita-server`: a revoked key marks the source unauthorized "with an
                // explanation and an action to enter a new key". The key is still in the
                // keystore, so a page can be built, so without this the browser opened and
                // every request in it failed in silence.
                source.state is SourceConnectionState.Unauthorized ->
                    navigation.push(Screen.SourceRefused(source))
                else -> {
                    val credentials = dependencies.credentials
                    val page: Screen? =
                        CataloguePage.of(source, credentials)?.let { Screen.Catalogue(it) }
                            ?: KavitaPage.of(source, credentials)
                                ?.let { Screen.Kavita(it, KavitaLevel.Libraries, term) }
                            ?: SmbPage.of(source, credentials)?.let { Screen.Share(it) }
                    if (page == null) navigation else navigation.push(page)
                }
            }
        },
        sheet = { sheet = it },
    )

    AppIntents(
        host = host,
        settings = settings,
        handedOver = handedOver,
        quickAction = quickAction,
        isReading = navigation.current is Screen.Reader,
        onRefusedFile = { refusedFile = it },
    )

    refusedFile?.let { outcome ->
        // Named, not swallowed. `local-library`: the app "names the format it detected and
        // states which formats it supports, rather than reporting a generic failure".
        RefusedFileDialog(outcome = outcome, onDismiss = { refusedFile = null })
    }

    AdaptiveNavigationShell(
        entries = AppDestination.entries.map { destination ->
            NavigationEntry(
                label = stringResource(destination.label),
                icon = destination.icon,
                selected = navigation.destination == destination,
                onSelect = { navigation = navigation.select(destination) },
            )
        },
        menu = RailMenuLabels(
            expand = stringResource(R.string.rail_expand),
            collapse = stringResource(R.string.rail_collapse),
        ),
        // Only the opened rail draws these. The library's own top bar carries Shelves and
        // Settings on a narrow window and deliberately drops both once the window is wide
        // enough for side navigation — so without them here, a tablet reader could reach
        // neither.
        secondaryEntries = listOf(
            NavigationEntry(
                label = stringResource(R.string.destination_shelves),
                icon = Icons.Filled.Inventory2,
                selected = navigation.current is Screen.Shelves,
                onSelect = {
                    // Choosing the entry a reader is already on is not a second copy of it.
                    if (navigation.current !is Screen.Shelves) {
                        navigation = navigation.push(Screen.Shelves)
                    }
                },
            ),
            NavigationEntry(
                label = stringResource(R.string.destination_settings),
                icon = Icons.Filled.Settings,
                // Never selected: Settings is a screen the reader comes back from rather
                // than a place they stay in, and an indicator left on it would be claiming
                // the library was somewhere else.
                selected = false,
                onSelect = {
                    downloads.value = dependencies.downloads.library()
                    navigation = navigation.push(Screen.Settings())
                },
            ),
        ),
        showsNavigation = navigation.showsNavigation,
    ) {
        // `native-experience` asks for predictive back, and the manifest opt-in is only the
        // half the system can do for itself: it draws the way out of the *app*. Everything
        // inside the app is this navigation value, so the preview of one of its screens
        // leaving is the app's own to draw — and until this host existed the handler below
        // was a plain `BackHandler` that reported only that the gesture had finished.
        //
        // Around the navigation area rather than around the whole shell, because back never
        // takes the navigation control away: it pops a screen off the current destination's
        // path, or returns to the destination the app opens on. The screen is what leaves;
        // the bar or the rail is what stays, and shrinking it with the screen would preview
        // an exit that is not going to happen.
        //
        // Settings keeps its own host inside this one. The composition local is static, so
        // the innermost host owns the handlers below it and each transform is applied
        // exactly once — nesting is the arrangement, not a conflict.
        PredictiveBackHost {
            // The one back rule. Registered before the content so that a modal or a screen
            // with its own handler wins over it, which is what those handlers are for;
            // everything else falls through to here.
            PredictiveBack(enabled = navigation.canGoBack) { navigation = navigation.back() }

            // One column, or a shelf and a page side by side where the window has room for
            // both. Which of the two is derived from this same navigation — see `AppPanes`.
            AppContent(
                host = host,
                navigation = navigation,
                settings = settings,
                onSettingsChange = onSettingsChange,
                onResetSettings = onResetSettings,
            )
        }
    }

    AppSheets(host = host, sheet = sheet)

    // Over whatever the reader landed on, and never blocking: one action dismisses it, the
    // scrim dismisses it, and back dismisses it. Nothing waits on it and nothing is lost by
    // ignoring it. A `ModalBottomSheet` rather than a full-screen dialog — see
    // `WhatsNewSheet`, which carries Material's own reason and the tablet it turns on.
    whatsNew?.let { release ->
        WhatsNewSheet(release = release, onDismiss = { whatsNew = null })
    }
}

/** What the navigation control calls each destination. Never an icon alone. */
private val AppDestination.label: Int
    get() = when (this) {
        AppDestination.HOME -> R.string.destination_home
        AppDestination.LIBRARY -> R.string.destination_library
        AppDestination.DOWNLOADS -> R.string.destination_downloads
        AppDestination.SEARCH -> R.string.destination_search
    }

private val AppDestination.icon: ImageVector
    get() = when (this) {
        AppDestination.HOME -> Icons.Filled.Home
        AppDestination.LIBRARY -> Icons.AutoMirrored.Filled.MenuBook
        AppDestination.DOWNLOADS -> Icons.Filled.Download
        AppDestination.SEARCH -> Icons.Filled.Search
    }

File: apps/ios/App/AppShell.swift
import SwiftUI

import EpubReaderFeature
import LibraryFeature
import Persistence
import Playback
import PlayerFeature
import SettingsFeature
import StoryArcCore

/// The four destinations, as the platform's own tab bar.
///
/// Until this existed the iPhone had **no tab bar at all**: `StoryArcApp` put `LibraryView`
/// straight into the `WindowGroup`, so the app was the shelf and everything else was
/// behind chrome on it. `navigation-shell` asks for destinations reachable at any
/// time from one persistent control, and for that control not to grow a row when a reader
/// adds a server. This is that control.
///
/// ## The search role was here, and was removed
///
/// **The argument this comment used to make**, at the head of the list below: *"`Tab(role:
/// .search)` is set apart from the three rather than listed among them — the circular
/// button on the trailing edge — and expands into a field that takes the rest of the bar
/// with it. That is the requirement `navigation-shell` states as 'set apart from them
/// rather than listed among them', and it is one argument."*
///
/// **Why it no longer holds.** The argument was sound about the requirement it quoted, and
/// that requirement is the thing that changed. It named a *control* — "set apart from" —
/// and the control it got does more than sit apart: it morphs the tab into a text field in
/// place, confirmed on a device on 2026-08-31. So the bar changed shape under the reader's
/// thumb, and because there was no screen behind it there was nowhere to land and nothing
/// the app could offer before a letter was typed. `navigation-shell` is now written as an
/// outcome instead: search "SHALL be a place a reader arrives at, and no control SHALL
/// change shape or position to become it".
///
/// **What removing the role costs.** Placement on the trailing edge, the circular
/// treatment, and the morph — all three, and they were genuinely the platform's own work.
/// What it buys is the thing the requirement now asks for: a destination with a screen
/// behind it, drawn like its three neighbours, that leaves the control the reader's thumb
/// is resting on exactly where it was. See ``LibraryDestination`` for why *this* app in
/// particular needs search to be a place rather than a filter.
///
/// **The bar keeps its floating capsule**, which is what iOS 26 draws. Android's
/// deliberately does not — `ShortNavigationBar` exposes no `shape` parameter at all, so a
/// capsule is inexpressible there rather than merely discouraged. That divergence is in the
/// change's `design.md` and is [ADR-0001](../../docs/decisions/0001-independent-native-cores.md)
/// working as intended.
///
/// Three things the system still gives us here that hand-building would not:
///
/// - **`.tabViewStyle(.sidebarAdaptable)`** makes the same set the iPad's sidebar without a
///   second navigation to keep in step, and the adaptation is non-destructive: rotating
///   back does not lose the destination, its scroll position or its filters.
/// - **`.tabBarMinimizeBehavior(.onScrollDown)`** is *"chrome that gets out of the way"* —
///   the bar recedes as covers scroll and comes back on the way up, with no animation when
///   the system asks for reduced motion, and no state in which it cannot be recovered.
/// - **A safe area.** The floating search pill this replaces had none, so cover titles at
///   the foot of the grid rendered *behind* it. A real tab bar insets its content.
///
/// The docked-transport slot below the tabs — `tabViewBottomAccessory`, the mini player in
/// the reference the owner supplied — held this comment's predecessor open and empty for a
/// reason it stated honestly: the only transport this app had was EPUB read-aloud, it lived
/// inside a reader presented as a full-screen cover, and speech ended when that cover was
/// dismissed, so there was no navigation behind it to dock to. `read-aloud-beyond-the-reader`
/// moved the session out of the screen, and `audiobooks-and-playback` moved it again — into
/// `PlayerCentre`, which a narrated audiobook and a synthesised voice both drive. So the slot
/// carries one `PlayerDock` for both. Nothing else is put at that edge; it is the app's one
/// persistent transport, because the platform already offers the rest on the lock screen.
struct AppShell: View {
    /// What a tab is worth as a selection.
    ///
    /// All four destinations come from ``LibraryDestination`` rather than being restated,
    /// so the promise that the set does not grow with what a reader configures is held in
    /// one place and tested there. Search had a `case` of its own here while it was a role,
    /// because a role is not a destination and the shell still had to address its tab; now
    /// that it is a destination, the extra case would be a second way to say the same thing
    /// and a second thing to keep in step.
    enum Selection: Hashable {
        case destination(LibraryDestination)
        /// A row the iPad sidebar reveals below the four. Not a destination and never in
        /// the tab bar — see ``SidebarEntry`` and ``LibrarySidebar``.
        case sidebar(SidebarEntry)
    }

    /// Where the reader is. `navigation-shell`: the app opens on the home surface, unless
    /// the launch named somewhere else.
    @Binding var tab: Selection

    /// What changed in the version just installed, if this launch is the one that says so.
    ///
    /// `settings-and-about`: shown "once, after it has been updated", never on a first ever
    /// launch, and never twice for one version. All three are ``WhatsNew/onLaunch(store:)``'s
    /// answer, and it is asked **here, in the initial value of the state** rather than in a
    /// `.task`: the version is recorded by the same call that decides, so a reader who swipes
    /// the sheet away — or one who never sees it because the version had nothing to say — is
    /// recorded either way. There is no path through this that shows the screen without
    /// recording it, which is the requirement stated as "the seen flag is written when the
    /// screen is shown, not when it is dismissed".
    ///
    /// It is the shell's rather than `StoryArcApp`'s because this is what a reader lands on,
    /// and `design.md` puts the presentation here.
    @State private var whatsNew: WhatsNewRelease? = WhatsNew.onLaunch()
    /// Whether the full player is presented.
    ///
    /// **Held here, by the shell, because the shell is what outlives a session.** It was
    /// `@State` inside `PlayerDock`, which attached the presentation inside
    /// `if let bar = centre.compact` — a computed value that changes on every play/pause and
    /// every chapter boundary, so pressing a transport control inside the player destroyed the
    /// player's own host. `PlayerSheet.swift` in `PlayerFeature` carries the full account.
    ///
    /// The dock still decides *when* to open it; it no longer hosts it.
    @State private var isShowingPlayer = false

    let model: LibraryModel
    let progress: ProgressStore?
    let onOpen: (Publication, URL) -> Void
    let onOpenSettings: () -> Void
    /// See ``LibraryView/init(model:surface:progress:onOpen:showLibrary:)``.
    let showLibrary: Int

    var body: some View {
        // Read in `body`, where Observation registers the dependency, rather than inside
        // the accessory's own builder, which SwiftUI may run later. The narrow question —
        // is a session running — and not the book, whose chapter is rewritten on every
        // sentence: the shell has no business redrawing three times a minute for hours.
        let isPlaying = PlayerCentre.shared.isRunning

        TabView(selection: $tab) {
            Tab(value: .destination(.home)) {
                HomeScreen(
                    model: model,
                    onOpen: onOpen,
                    onOpenSettings: onOpenSettings
                )
            } label: {
                label(Text("tab.home"), LibraryDestination.home.symbolName)
            }

            Tab(value: .destination(.library)) {
                library(.shelf)
            } label: {
                label(Text("tab.library"), LibraryDestination.library.symbolName)
            }

            // Not `library(.onDevice)` any more. That drew the right *set* of covers and
            // nothing else — no queue, no idea what the files weigh, no way to remove one
            // — with the rest of it three taps inside the Settings modal. The destination
            // owns all of it now; see ``DownloadsDestination``.
            Tab(value: .destination(.onDevice)) {
                DownloadsDestination(
                    model: model,
                    onOpen: onOpen,
                    onShowLibrary: { tab = .destination(.library) }
                )
            } label: {
                label(Text("tab.downloads"), LibraryDestination.onDevice.symbolName)
            }

            // **No `role:`.** A fourth tab, drawn like its three neighbours, leading to a
            // page — see this type's own comment for what the role did and why it went. The
            // label was always ours and stays ours: the system would write its own in the
            // *device's* language, and `localization` lets a reader choose the app's without
            // touching the device's.
            Tab(value: .destination(.search)) {
                library(.search)
            } label: {
                label(Text("tab.search"), LibraryDestination.search.symbolName)
            }

            // The iPad's second half of the same set: library sections and the reader's
            // shelves, under their own headers, hidden from the tab bar so the phone still
            // shows four destinations and nothing else.
            LibrarySidebar(model: model, onOpen: onOpen) { Selection.sidebar($0) }
        }
        .tabViewStyle(.sidebarAdaptable)
        .tabBarMinimizeBehavior(.onScrollDown)
        // The docked transport, for everything that speaks. `audio-playback` requires that
        // it reserve no space when there is no session, and ``PlaybackAccessory`` is what
        // holds that promise — see its own comment for why an empty builder is not enough.
        .modifier(
            PlaybackAccessory(
                isPlaying: isPlaying,
                isShowingPlayer: $isShowingPlayer,
                onReturn: onOpen
            )
        )
        // The player itself, hosted here rather than in the accessory above.
        //
        // `audio-playback` asks that opening the player "never restarts, reloads or
        // repositions the audio", and the accessory could not keep that promise: the bar reads
        // the book's chapter, so the bar is rebuilt as the audio moves, and a presentation
        // attached inside it was torn down by its own transport controls. This is attached to
        // the `TabView`, which exists for as long as the app does.
        .playerSheet(isPresented: $isShowingPlayer, centre: .shared)
        // What changed, once, over whatever the reader landed on.
        //
        // `.large` and one detent, because the content is a heading and four rows: a medium
        // detent would show two of them and a drag handle, which reads as a card the reader
        // has to open rather than a thing the app is telling them. `design.md` settles it,
        // and Android's `ModalBottomSheet` is capped and expandable for the reason Material
        // gives there — the two platforms are deliberately not the same shape here.
        //
        // Never blocking: dismissing it needs one action and the tab bar is behind it, not
        // gone. Nothing waits on it and nothing is lost by ignoring it.
        .sheet(item: $whatsNew) { release in
            WhatsNewSheet(release: release) { whatsNew = nil }
                .presentationDetents([.large])
        }
        // `navigation-shell`: leaving search returns the reader to the destination they
        // were on "with its scroll position and filters intact". The query narrows the one
        // library, so a term left behind would follow them onto the shelf and leave it
        // looking half-empty for no reason they could see. The term itself is not lost —
        // the model has already filed it as a recent search.
        .onChange(of: tab) { previous, _ in
            if previous == .destination(.search) { model.query.search = "" }
        }
        // The library is brought up here rather than by whichever surface happens to be
        // shown first. It used to be started by the shelf's own `.task`, which was
        // correct while the shelf *was* the app — and became a bug the moment a reader
        // could land somewhere else: home opened onto an empty library until they had
        // visited the library tab at least once. `home-screen` says home is assembled
        // from what the device already knows, so what the device knows has to be read
        // before home is drawn.
        .task {
            model.restoreFolders()
            await model.refreshProgress()
        }
    }

    /// One tab's label. Named so the four of them cannot drift apart.
    private func label(_ title: Text, _ symbol: String) -> some View {
        Label {
            title
        } icon: {
            Image(systemName: symbol)
        }
    }

    /// The library, on one of its three faces.
    ///
    /// The same view each time, and deliberately: the shelf, the on-device shelf and
    /// search are one screen over three sets, so a cover looks and behaves the same on all
    /// of them and opens the same reader.
    private func library(_ surface: LibrarySurface) -> some View {
        LibraryView(
            model: model,
            surface: surface,
            progress: progress,
            onOpen: onOpen,
            showLibrary: showLibrary
        )
    }
}

/// The slot below the tabs, present only while a voice is speaking.
///
/// **An empty builder is not an absent accessory.** The first attempt passed
/// `tabViewBottomAccessory { if isSpeaking { ... } }`, on the reasoning that producing no
/// content leaves the slot nothing to make room for. A screenshot of the library with no
/// session settled it: the platform draws the glass capsule regardless, so the shelf
/// gained an empty pill above the tab bar and every destination lost that much height.
/// The comment left behind said this was the outcome to watch for and named the remedy,
/// which is the only reason this took one capture rather than an afternoon.
///
/// `tabViewBottomAccessory(isEnabled:)` is that remedy and it is **iOS 26.1** against this
/// app's 26.0 floor ([ADR-0003](../../docs/decisions/0003-platform-floors.md)), so it costs
/// the availability branch below — the app's only one. On 26.0 the empty capsule remains:
/// the old behaviour, not a new defect, and the honest cost of a floor set before the API
/// existed. Delete the branch when the floor moves.
private struct PlaybackAccessory: ViewModifier {
    let isPlaying: Bool
    /// Passed through to the bar, which sets it. The presentation it drives is attached to the
    /// `TabView` above, not inside this accessory — see `PlayerSheet.swift`.
    @Binding var isShowingPlayer: Bool
    let onReturn: (Publication, URL) -> Void

    func body(content: Content) -> some View {
        if #available(iOS 26.1, *) {
            content.tabViewBottomAccessory(isEnabled: isPlaying) { bar }
        } else {
            content.tabViewBottomAccessory {
                if isPlaying { bar }
            }
        }
    }

    /// One slot, one bar, and one session behind it.
    ///
    /// **There were two bars here until `audiobooks-and-playback` §4.2, and the reason is
    /// worth keeping.** A narrated audiobook took the slot through `PlayerDock` and a
    /// synthesised voice took it through a `ReadAloudDock` of its own, because a read-aloud
    /// session driving `PlayerCentre` would have offered a speed control that did nothing —
    /// Readium 3.11.0 sets no rate on an utterance — and `audio-playback` forbids a control
    /// that is "present and refusing". `SpeechRate` and `SpokenVoice` answered that through
    /// the delegate Readium points the caller at, so the second bar went: `design.md` asks
    /// for one session object behind both sources, and two surfaces drawing it was the last
    /// place the two could have disagreed.
    ///
    /// The way back is `onReturn` — the same seam the shelf uses to open a cover, taking the
    /// publication and its URL. Only a publication being read aloud uses it; a narrated
    /// audiobook has no screen to go back to and its row opens the player instead. Opening
    /// the book that is already being spoken is what `SessionHandover` answers with `adopt`:
    /// the reader picks up the sentence the voice is on and the voice never notices.
    private var bar: some View {
        PlayerDock(
            centre: PlayerCentre.shared,
            isShowingPlayer: $isShowingPlayer,
            onReturn: onReturn
        )
    }
}
